components/ai_chat/resources/page/components/input_box/ranges.ts

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2025 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+import type { ContentNode } from "./editable"
+
+export const createNode = (node: ContentNode) => {
+  if (typeof node === 'string') {
+    return document.createTextNode(node)
+  }
+
+  // Ideally we'd use a leo-label here, but shadowRoot does not play nice with
+  // contenteditable.
+  if (node.type === 'associated-content') {
+    const el = document.createElement('span')
+    el.contentEditable = 'false'
+    el.dataset.text = node.text
+    el.dataset.type = node.type
+    el.dataset.id = node.id
+    el.textContent = `@${node.text}`
+
+    return el
+  }
+
+  throw new Error('Unknown content type: ' + JSON.stringify(node))
+}
+
+export const findParentEditable = (node: Node | null): HTMLElement | undefined => {
+  if (node instanceof HTMLElement && node.dataset.editor) {
+    return node
+  }
+
+  return node?.parentNode ? findParentEditable(node.parentNode) : undefined
+}
+
+export const selectRange = (range: Range) => {
+  const selection = window.getSelection()
+  if (!selection) return
+
+  selection.removeAllRanges()
+  selection.addRange(range)
+}
+
+export const insertAtCursor = (contentNode: ContentNode) => {
+  const selection = window.getSelection()
+  if (!selection) return
+  if (!findParentEditable(selection.anchorNode)) return
+
+  const node = createNode(contentNode)
+
+  const range = selection.getRangeAt(0)
+  range.insertNode(node)
+
+  // Make sure the node isn't selected by collapsing the range
+  // to the end.
+  range.collapse()
+}
+
+export const getRangeToTriggerChar = (triggerChar: string) => {
+  const selection = window.getSelection()
+  if (!selection) return
+  if (!findParentEditable(selection.anchorNode)) return
```

Comment:
What if there are more than 1 `Editable`s on the page. It's not a problem because we expect the caller to only call this function when their own content editable has selection? Could we pass a ref to the Editable in?

Reply:
Yeah, at the moment I've been only expecting it to be called when your own editable has the selection. I'll have a bit of a think about this and see if I can come up with something nicer

Reply:
updated with an API you pass what you want to edit into - its not perfect but I figure it'll be easy to adapt if we ever do have multiple inputs